simexpal/base.py

```diff
@@ -693,72 +693,191 @@ def check_available(self):
+				cmd = [util.expand_at_params(arg_tmpl, substitute) for arg_tmpl
+						in self._inst_yml['generator']['args']]
+
+				partial_path = os.path.join(instance_dir, self.unique_filename)
+				with open(partial_path + '.gen', 'w') as f:
+					subprocess.check_call(cmd, cwd=self.config.basedir,
+							stdout=f, stderr=subprocess.PIPE)
+				os.rename(partial_path + '.gen', partial_path + '.post0')
+			elif 'method' in self._inst_yml:
+				print(f"Downloading instance '{self.shortname}' with method '{self.method}'")
+
+				instances.download_instance(self._inst_yml,	instance_dir, self.unique_filename,
+					os.path.join(instance_dir, self.unique_filename), '.post0')
+			else:
+				raise RuntimeError(f"Unknown installation option for instance '{self.shortname}'")
 
 		stage = 0
+		filenames = self.filenames
 		if 'postprocess' in self._inst_yml:
+
 			did_work = False
-			if not os.path.isfile(partial_path + '.postprocessed'):
-				assert self._inst_yml['postprocess'] == 'to_edgelist'
-				instances.convert_to_edgelist(self._inst_yml,
-						partial_path + '.post0', partial_path + '.post1')
-				os.rename(partial_path + '.post0', partial_path + '.original')
-				stage = 1
-				did_work = True
-				util.touch(partial_path + '.postprocessed')
+			postprocessed = os.path.isfile(os.path.join(instance_dir, self.yml_name) + '.postprocessed')
+			if not postprocessed:
+
+				# Make sure that the .post0 files exist.
+				# For cases like 'repo: local' instances, adding the 'postprocess' key
+				# after downloading the instance or re-postprocessing instances.
+				for file in filenames:
+					if not os.path.isfile(os.path.join(instance_dir, file + '.post0')):
+						os.rename(os.path.join(instance_dir, file), os.path.join(instance_dir, file + '.post0'))
+
+				print("simexpal: Start postprocessing of instance '{}'".format(self.shortname))
+				if isinstance(self._inst_yml['postprocess'], str):
+					assert self._inst_yml['postprocess'] == 'to_edgelist'
+
+					partial_path = os.path.join(instance_dir, self.unique_filename)
+					try:
+						print("simexpal: Converting instance '{}' to edge list format".format(self.shortname))
+						instances.convert_to_edgelist(self._inst_yml, partial_path + '.post0', partial_path + '.post1')
+						os.rename(partial_path + '.post0', partial_path + '.original')
+					except RuntimeError as e:
+						print("simexpal: An error occurred while postprocessing the instance '{}': {}".format(
+							self.shortname, e))
+						print("simexpal: Restoring original instance files")
+						os.rename(partial_path + '.post0', partial_path)
+						print("simexpal: Skipping postprocessing of instance '{}'".format(self.shortname))
+						return
+
+					stage = 1
+					util.touch(os.path.join(instance_dir, self.yml_name) + '.postprocessed')
+					did_work = True
+				else:
+					assert isinstance(self._inst_yml['postprocess'], list)
+
+					import subprocess
+					import shutil
```

Comment:
The indentation level here is quite deep. Can we extract this code into a function?

Reply:
You mean the code from line ``789`` to ``869``?

Reply:
I think the line numbers changed due to rebasing. Could you specify which lines we should extract into a function?